includes/abstracts/abstract-wc-stripe-payment-gateway.php

```diff
@@ -647,24 +664,29 @@ public function prepare_source( $user_id, $force_save_source = false, $existing_
 
 			$source_id = $wc_token->get_token();
 
-			if ( $this->is_type_legacy_card( $source_id ) ) {
-				$is_token = true;
+			if ( ! $this->is_type_legacy_card( $source_id ) ) {
+				$source_object = self::get_source_object( $source_id );
+			}
+		} else {
+			// New CC info was entered and we have a new source to process.
+			if ( ! empty( $_POST['stripe_source'] ) ) {
+				$source_object = self::get_source_object( wc_clean( $_POST['stripe_source'] ) );
+				$source_id     = $source_object->id;
+			} elseif ( isset( $_POST['stripe_token'] ) && 'new' !== $_POST['stripe_token'] ) {
+				$source_id = wc_clean( $_POST['stripe_token'] );
 			}
-		} elseif ( isset( $_POST['stripe_token'] ) && 'new' !== $_POST['stripe_token'] ) {
-			$stripe_token     = wc_clean( $_POST['stripe_token'] );
-			$maybe_saved_card = isset( $_POST[ 'wc-' . $payment_method . '-new-payment-method' ] ) && ! empty( $_POST[ 'wc-' . $payment_method . '-new-payment-method' ] );
 
-			// This is true if the user wants to store the card to their account.
-			if ( ( $user_id && $this->saved_cards && $maybe_saved_card ) || $force_save_source ) {
-				$response = $customer->add_source( $stripe_token );
+			if ( $this->should_save_payment_method( $source_object, $customer, $force_save_source ) ) {
+				$response = $customer->add_source( $source_id );
```

Comment:
Don't take this as a blocker for this PR, but have we (Hydra) ever had a conversation about migrating off of sources? I'll start a P2 thread 

Related: https://stripe.com/docs/payments/payment-intents/migration-synchronous and especially https://stripe.com/docs/payments/payment-methods/transitioning